### DISCUSSION TITLE
How to properly remove mods from mod list?

### DISCUSSION THREAD

**Discussion by hodge47** [Hello-Meow/ModTool#19] [Q&A] [answered]
Hello,
I am struggling to remove mods from the mods list in the ModManager.

Some back story - I'm using mod.io to host mods that are downloaded/updated/removed from an in-game browser. Unfortunately, when a mod is updated, the path to the mod changes, and likewise with deleting mods, the path is non-existent. 

When the mod browser is closed, I try to update the ModManager with the new or deleted paths via `ModManager.RemoveSearchDirectories()`. Before calling this, I unload all mods/scenes and iterate through a list of mod paths and check for directory existence. Once the non-existing paths are passed to  `ModManager.RemoveSearchDirectories()` I call `ModManager.Refresh()` before collecting all the installed mods. 

I have a list of available mods that gets added/removed from with `ModManager.ModFound` and `ModManager.ModRemoved` events, but `ModManager.ModRemoved` never seems to fire. This list is then used in my UI to select levels, but I am getting duplicates because mods are not being removed.
I am getting a conflicting scene warning so maybe the mod is not being removed at all, just updated in ModManager?
![image](https://user-images.githubusercontent.com/11467129/212772942-615a4b70-66ba-47b7-b270-97461324ada1.png)

Is there something that I am forgetting to do?

Attempt to refresh mod list:
 ```csharp
private IEnumerator IRefreshModManager(Action afterAction = null)
        {
            yield return null;
            
            // Unload all mods
            foreach (var mod in ModManager.mods)
            {
                foreach (var scene in mod.scenes)
                {
                    scene.Unload();
                }
                mod.Unload();
            }

            // Remove any dangling modio search directories
            foreach (var directory in _modDirectories)
            {
                if (!Directory.Exists(directory))
                {
                    ModManager.RemoveSearchDirectory(directory);
                }
            }
            
            // Refresh the mod manager
            ModManager.Refresh();

          .............
}
```

ModManager.ModRemoved event subscription in Start():
```csharp
ModManager.ModRemoved += OnModRemoved;
```

OnModRemoved method:
```csharp
private void OnModRemoved(Mod mod)
        {
            Debug.Log($"{mod.modInfo.name} was removed...");
            if (_availableMods.Contains(mod))
            {
                _availableMods.Remove(mod);
            }
        }
```

**Comment by Hello-Meow** [comment, marked as answer]
I'm not familiar with modio. How and where does modio store mods?

Search directories should be any folders than can contain mods, not individual mod folders themselves. Ideally you would only set these up one time at the start. I'd only add directories than can contain mods and then use ModManager.Refresh() to update the mods after a mod has been added or removed.

When a mod is unloaded, all the scenes will be unloaded first, so you don't have to do that separately. Unloading mods can take more than one frame. For example, if you try to unload a mod when a scene is still loading, it has to wait for it to load before it can unload everything. This probably isn't an issue here, but you shouldn't expect mods to be unloaded right away in every case.

**Reply by hodge47** [reply]
Thank you, that was my problem! I guess I didn't pay enough attention to realize all the directories were searched recursively to find the mods. 

This is the current path to one of my installed mods from modio - `<Application.PersistentDataPath>\mod.io\03761\data\mods\2338493_3423860\First Ever Mod Track`
I'm guessing the 03761 is my user ID number for modio, but after adding <Application.PersistentDataPath>\mod.io to the directory to search it started working as intended and mods were getting updated and removed properly in the ModManager after an action occurred with modio browser.

As for unloading mods, that was just there as a test, I'm actually unloading them when returning to the main menu using `Mod.Unload()`. I was getting a little desperate yesterday 🤣

**Reply by Hello-Meow** [reply]
Good to hear!
